araig_interpreters/src/location_interpreters/vicon_interpreter.cpp

```diff
@@ -0,0 +1,135 @@
+#include <ros/ros.h>
+#include <tf/transform_broadcaster.h>
+#include <tf/transform_listener.h>
+#include <geometry_msgs/TransformStamped.h>
+#include <geometry_msgs/PoseStamped.h>
+#include <std_msgs/String.h>
+#include <ros/master.h>
+#include <vector>
+#include <map>
+#include <string>
+#include <boost/algorithm/string.hpp>
+#include <boost/bind.hpp>
+
+class vicon_interpreter
+{
+private:
+    ros::NodeHandle nh;
+    ros::NodeHandle nh_priv;
+
+    std::vector<std::string> viconTopics;
+    ros::master::V_TopicInfo viconTopicInfos;
+
+    tf::TransformListener tf_ls;
+    tf::Transform trans_ls;
+
+    geometry_msgs::PoseStamped pose_msg;
+
+    std::map<std::string, ros::Publisher> viconPubList;
+    std::map<std::string, ros::Subscriber> viconSubList;
+
+    std::map<std::string, geometry_msgs::PoseStamped> OutputPoseMsg;
+
+public:
+
+    void stampedTrans2Pose(geometry_msgs::TransformStamped transMsg, 
+        geometry_msgs::PoseStamped& poseMsg){
+            
+            poseMsg.pose.position.x = transMsg.transform.translation.x;
+            poseMsg.pose.position.y = transMsg.transform.translation.y;
+            poseMsg.pose.position.z = transMsg.transform.translation.z;
+
+            poseMsg.pose.orientation.w = transMsg.transform.rotation.w;
+            poseMsg.pose.orientation.x = transMsg.transform.rotation.x;
+            poseMsg.pose.orientation.y = transMsg.transform.rotation.y;
+            poseMsg.pose.orientation.z = transMsg.transform.rotation.z;
+
+            poseMsg.header = transMsg.header;
+    }
+
+    void doTransform(geometry_msgs::PoseStamped& pose_msg, geometry_msgs::PoseStamped& pose_msg_tf)
+    {
+        ros::Time now = ros::Time(0);
+        try{
+            tf_ls.waitForTransform("/map", "/vicon/world", now, ros::Duration(1.0));
+            tf_ls.transformPose("/map", pose_msg, pose_msg_tf);
+        }catch(tf::TransformException ex){
+            ROS_ERROR("%s", ex.what());
+        }
+    }
+
+    void getViconTopic(ros::master::V_TopicInfo& topicInfos, std::vector<std::string>& viconTopics)
+    {
+        std::vector<std::string> lv_elems;
+        for(ros::master::V_TopicInfo::iterator it = topicInfos.begin();
+            it != topicInfos.end(); it++){
+                boost::algorithm::split( lv_elems, it->name, boost::algorithm::is_any_of( "/" ) );
+                ROS_INFO_STREAM("examining topic "<<it->name);
+
+                std::string topicName = "";
+                for(uint i=0; i<lv_elems.size(); ++i){
```

Comment:
wait why do you split the topic name above, and then reassemble it here?

Reply:
In new version, using string::find to substitute these